### ISSUE TITLE
[request] [help] change default theme background without recompile

### ISSUE THREAD

**Issue by Miosame** [AnySoftKeyboard/AnySoftKeyboard#2470]
First of all thanks for this amazing project!

I've seen people changing backgrounds easily for the "yochees dark" theme, but nobody ever mentions the way they did it. I couldn't find a way to do it in none of the theme or tweak settings, so I assume one has to recompile the base apk by changing the background definition for that theme and then recompile for every single apk release? Is there any easier way to keep up with upstream changes instead while keeping a custom background? Thanks!

**Comment by lubenard** [comment]
Hello, i am not sure what you are talking about.

Have you looked in ```Interface -> Pick and configure keyboard themes``` ?

Are you looking for something different ?

**Comment by Miosame** [comment]
@lubenard indeed I am searching for adding a background image behind the keys, here's a swiftkey example:

![image](https://user-images.githubusercontent.com/8201077/93019704-f9bee600-f5d8-11ea-9999-c92ae7492079.png)

`Interface -> Pick and configure keyboard themes` does not have such option available.

**Comment by lubenard** [comment]
I might be wrong, but you cannot add custom image as keyboard background.

I you want to developp that feature, it would be awesome !

**Comment by Miosame** [comment]
@lubenard I sadly have barely any experience with native android development, so I can't add it, best I could think of was to attempt to copy over all resources from the "yochees dark" theme into one of those: https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/master/addons/themes/PACKS.md - to be independent of upstream anysoftkeyboard changes - the israel theme being the only one also that has a background image, but has some layout issues from what I understand.

**Comment by lubenard** [comment]
Yes, i think creating a theme is the best thing you can do !
There is still a lot to do on this keyboard, but for now this is your best option

**Comment by Miosame** [comment]
@lubenard it would really help if there was any resources for people how to actually make themes, since it's mostly just figuring it out from examples for sure anyway.

**Comment by lubenard** [comment]
Yes, i am not into theme making, but you can ask menny about it.

There is a doc for theme [here](https://github.com/AnySoftKeyboard/AnySoftKeyboard/blob/master/addons/CONTRIBUTING.md)

For your error, do you have Java 11 ? It is a typical error when using different version of java.

Please make sure you meet those requirements:

- Java 11 - get OpenJDK or AdoptOpenJDK. If you are using jEnv then we have a .java-version file already set up.
- Android Studio version 4.0.
      You will need to point your Android-Studio to use the installed Java11. You can configure the path inside Android Studio menu: File > Project Structure > JDK Location.
- Android SDK API Level 28 installed.
        Set up local.properties file to point to it (that is, ensure you have sdk.dir=/path/to/android/sdk in it).
    Android NDK r14b installed. Set up local.properties file to point to it (that is, ensure you have ndk.dir=/path/to/android/ndk in it).

**Comment by Miosame** [comment]
I'll check the rest, but I think I might be failing Java 11 requirements already?

```
# java -version
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (build 1.8.0_265-b01)
OpenJDK 64-Bit Server VM (build 25.265-b01, mixed mode)
```
The docs don't seem to mention themes at all, but some of that info I hope is transferable, thanks!

> Yes, i am not into theme making, but you can ask menny about it.
@menny is there theme related starter docs?

**Comment by menny** [comment]
This is a frequently requested feature, and we'll do it at some point . But, idk when.. sorry.

Also, there aren't theme docs. We need to create those, too.

And, you're right, you Java version is wrong.
I would suggest installing Java 11 and jenv too

**Comment by Miosame** [comment]
I'm not trying to hurry you, no worries, I appreciate any contribution and obviously understand. Just want to find a way to do it until it's an upstream feature. :+1: 

I'll install java11 and see what else may be failing.

**Comment by lubenard** [comment]
> I'll install java11 and see what else may be failing. 

If you are under linux, check under ```/usr/lib/jvm```. 
That is where my java versions are installed. 
Then, go in Android Studio and check ```File -> Project Structure -> Sdk Location -> Jdk location``` is pointing to the right place

**Comment by Miosame** [comment]
@lubenard indeed I'm on linux, though even with host access the android studio flatpak can't seem to see anything in /usr/lib/jvm, not sure why yet

**Comment by Miosame** [comment]
@lubenard that worked, so is there a way to preview the keyboard via android emulator in android studio? or do I need to push the apk to my device each time after gradle builds it.

**Comment by lubenard** [comment]
Glad to hear ! 

Yes, you can push the app to your emulator without rebuilding it entirely.

The build you will get on this emulator in just a debug build, and it's not recommended to use it on your smartphone for everyday use.

In android Studio, select your virtual emulator, then click 'run app' (Map + F10 or the green triangle next to your emulator name).
It will launch your emulator and launch the app

**Comment by Miosame** [comment]
@lubenard thanks for taking so much time btw - I've got the emulator but gradle does not seem to push the application itself, do I need to manually move the apk onto the device after gradle runs through? (if so where's the apk as I can't locate a single .apk output after gradle runs successfully)

> The build you will get on this emulator in just a debug build, and it's not recommended to use it on your smartphone for everyday use.

That should be fine, considering I want to build a easy modifiable background theme for people to build, not add features to the keyboard itself (yet?).

**Comment by lubenard** [comment]
>thanks for taking so much time btw
Your welcome, i hope it hellp you !

> do I need to manually move the apk onto the device after gradle runs through? 

You should not need to, how are you compiling the app ?

> That should be fine, considering I want to build a easy modifiable background theme for people to build, not add features to the keyboard itself (yet?).

Yes, it should be fine then !

**Comment by Miosame** [comment]
@lubenard 

> You should not need to, how are you compiling the app ?

It auto detected gradlew for me, so it executes it on each run/build.

**Comment by lubenard** [comment]
Yes, but it should transfer it automatically. 

Are you tring to build only one theme/addon or the app itself ?

**Comment by Miosame** [comment]
@lubenard for now I am just trying to build the full keyboard, that's what I see:

![image](https://user-images.githubusercontent.com/8201077/93030729-87282780-f625-11ea-8e94-08dd78539f19.png)

the emulator is already open, but nothing happens on-screen, no install prompt, the app doesn't show up in appview nor does the keyboard install into keyboard settings, not sure what I am missing.

**Comment by lubenard** [comment]
The 'gradlew' thing is not normal. It shoud look to something like this:

![Screenshot_20200914_011820](https://user-images.githubusercontent.com/42534397/93031081-70370480-f628-11ea-9f54-0465a9b6e5e6.png)

**Comment by Miosame** [comment]
@lubenard I was afraid of that, because mine all have a red x on them:

![image](https://user-images.githubusercontent.com/8201077/93033515-dfb3f080-f636-11ea-9fb6-a14cba0a04ea.png)

However it did let me build and it installed to the android emulator just fine! for a final note, would you know how I could build the default included themes? e.g. israel64, attempting to just build it like the "app" just outputs:

![image](https://user-images.githubusercontent.com/8201077/93033897-71702d80-f638-11ea-9648-338e964d5992.png)

**Comment by lubenard** [comment]
Yes, click on the list next to the hammer, then, instead of selecting the theme, click 'edit configuration'.

From there, select the module you want to install, for example 'israel64', then instead of 'Default activity' in the Launch Options menu, select 'Nothing', click Apply, it should now work!

**Comment by Miosame** [comment]
@lubenard sweet! that worked like a charm, thanks so much for all your help!
I'll experiment some more and see if I run into some errors again, so for now I'll keep the issue open, especially because if people want to do a similar thing, they can find this single issue instead of going across multiple. :smile:

**Comment by darkknight73** [comment]
Hello everyone. I've the "Default activity not found" problem and I tried selecting "Nothing", but it didn't work for me. when I do that I get no errors but app doesn't get installed on the emulator.
I don't think it's because of my java version but I'm telling you anyways: "java version "20.0.1" 2023-04-18
Java(TM) SE Runtime Environment (build 20.0.1+9-29)".

any help would be appreciated.
